### PR TITLE
feat: Mega-AutoML benchmark infrastructure

### PR DESCRIPTION
**Pull request by w4nderlust**
## Summary

Full infrastructure for running the largest Ludwig AutoML experiment ever attempted — hundreds of datasets × ~100 configs each, with structured results, baselines, and a visualizer-ready dashboard.

**7,730 lines added across 25 files, zero dependencies on the old framework.**

### Core benchmark engine (`benchmark/`)

- **`db.py`** — `BenchmarkDB`: per-run Parquet storage + consolidated index; pure-pandas analytics (`best_per_dataset`, `combiner_win_rates`, `progress_summary`); DuckDB used lazily only for complex queries
- **`scheduler.py`** — `BenchmarkScheduler`: SQLite-backed job queue (survives restarts); priority ordering; automatic retry up to `max_attempts`; `run_sequential()` and `run_with_ray()` execution modes
- **`runner.py`** — `run_experiment()`: SIGALRM/threading.Timer timeout enforcement; OOM detection; NaN-loss guard; status values `done/failed/timeout/oom`
- **`dataset_registry.py`** — `DatasetRegistry`: JSON-persisted dataset catalogue with OpenML/Ludwig/Kaggle registration helpers
- **`baselines.py`** — XGBoost, LightGBM, AutoGluon baselines with early stopping; TabRepo precomputed results loader
- **`cost_estimator.py`** — GPU-hour and AWS spot cost projection before launching; Rich + plain-text output
- **`dashboard.py`** — live terminal progress view (Rich tables with plain-text fallback)
- **`exporter.py`** — `export_dashboard()`: structured JSON hierarchy for the visualizer

### Scripts

- **`scripts/prepare_benchmark.py`** — end-to-end pipeline: register datasets → quality check → target detection → generate configs → write JSONL; `--dry-run` / `--resume` support
- **`scripts/generate_configs.py`** — per-dataset JSONL config generation standalone CLI
- **`scripts/run_benchmark.py`** — **unified execution CLI** (20+ flags):
  - Dataset selection: `--registry`, `--openml-suite`, `--ludwig-builtins`, `--datasets`, `--priority-min`, `--source-filter`
  - Config generation: `--n-configs`, `--seed`, `--regenerate-configs`, `--skip-config-gen`
  - Execution: `--mode sequential|ray`, `--max-concurrent`, `--gpus-per-trial`, `--time-limit-per-job`, `--max-jobs`, `--max-attempts`, `--dry-run`
  - Baselines: `--run-baselines`, `--run-autogluon`
  - Export: `--export-dashboard`, `--dashboard-dir`, `--no-run-details`
  - Misc: `--live-dashboard`, `--estimate-cost`, `--log-level`
- **`scripts/kaggle_discovery/`** — scrape, filter, and download Kaggle datasets

### Dashboard

- **`dashboard/index.html`** — single-file vanilla JS/CSS SPA (no build step, no CDN):
  - **Datasets tab**: sortable/filterable table with score heat-map pills (green/yellow/red by column percentile)
  - **Dataset detail**: Ludwig vs XGBoost vs LightGBM baseline comparison bars, combiner breakdown, ranked runs table with Δ-from-best
  - **Configs tab**: sortable table of all unique model configs with win rates and mean rank
  - **Config detail**: cross-dataset performance table, rank distribution dot map
  - **Combiners tab**: CSS bar chart + stats table (win rate, P25/P75 scores)
  - **Runs tab**: aggregated run list with status filter pills
  - Hash-based routing (`#datasets`, `#dataset/name`, `#configs`, `#config/hash`, ...)
- **`dashboard/serve.py`** — Python HTTP server; auto-symlinks data dir; `--port` / `--data-dir` flags

### Dashboard data layout (from `exporter.py`)

```
data/
  summary.json           ← global stats, top combiners, generated_at
  datasets.json          ← per-dataset summary rows (list page)
  combiners.json         ← per-combiner aggregate stats
  configs.json           ← per-config aggregate stats
  datasets/{name}.json   ← full ranked run list + baseline comparison
  configs/{hash}.json    ← cross-dataset performance profile
  runs/{run_id}.json     ← individual run detail
```

### Quick start

```bash
# 1. Register datasets and generate configs
python scripts/prepare_benchmark.py --openml-suite 99 \
    --registry benchmark/registry.json --configs-dir benchmark/configs

# 2. Run (sequential, 30 min/job, export dashboard on completion)
python scripts/run_benchmark.py \
    --registry benchmark/registry.json \
    --configs-dir benchmark/configs \
    --run-baselines --export-dashboard --live-dashboard

# 3. View results
python dashboard/serve.py --data-dir ./dashboard/data
```

## Test plan

- [ ] `tests/test_exporter.py` — 13 tests covering all export paths (empty DB, NaN values, baseline exclusion, registry metadata injection)
- [ ] `scripts/run_benchmark.py --dry-run --ludwig-builtins` — smoke test shows banner with correct job count
- [ ] `scripts/run_benchmark.py --estimate-cost` — prints cost table without training
- [ ] Manual: `dashboard/serve.py` + open browser, verify all tabs render and routing works

### CODE REVIEW

**Comment by w4nderlust** [comment]
Closing — moving to a fork under ludwig-ai to preserve the original HazyResearch repo.
